### DISCUSSION TITLE
Developing in VS Code

### DISCUSSION THREAD

**Discussion by khaeru** [iiasa/message_ix#672] [General]
Per some oral discussion at today's MESSAGEix team meeting, here's a thread to collect tips for and experiences **developing for MESSAGEix in Visual Studio Code**. In particular, we can talk about how to easily implement the [code style](https://docs.messageix.org/en/stable/contributing.html#code-style) that applies to `ixmp`, `message_ix`, `message-ix-models`, and `message_data` (private).

**Comment by khaeru** [comment]
Here are the extensions I use:
- [Python (`ms-python.python`)](https://marketplace.visualstudio.com/items?itemName=ms-python.python):
  - I use the open-source "jedi" language server rather than the closed-source Pylance.
- [Black Formatter (`ms-python.black-formatter`)](https://marketplace.visualstudio.com/items?itemName=ms-python.black-formatter), configured per the README to format on save.
- [isort (`ms-python.isort`)](https://marketplace.visualstudio.com/items?itemName=ms-python.isort), ditto.

Black output is compliant with flake8, so I don't use any separate tool for that. For typing with mypy, I instead use a Git hook, which I'll post about in a separate discussion/thread.

**Reply by gidden** [reply]
Amazing, thank you @khaeru ! Will look forward to the mypy hook.

**Reply by khaeru** [reply]
See [here](https://github.com/iiasa/message_ix/discussions/673#discussioncomment-4166040)—in a separate discussion because that's not specific to VS Code.

**Comment by khaeru** [comment]
An update: I've personally:
- Switched to using the [Ruff (`charliermarsh.ruff`)](https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff) extension to replace both Black and isort; I've uninstalled both.
- Added the first-party [Mypy (`ms-python.mypy-type-checker`)](https://marketplace.visualstudio.com/items?itemName=ms-python.mypy-type-checker) extension.

These two (plus `ms-python.python` with Jedi per [my earlier comment](https://github.com/iiasa/message_ix/discussions/672#discussioncomment-4165868)) are sufficient to implement the code style.

For convenience, I also make use of:
- [reStructuredText Syntax highlighting (`trond-snekvik.simple-rst`)](https://marketplace.visualstudio.com/items?itemName=trond-snekvik.simple-rst).
- [Todo Tree (`Gruntfuggly.todo-tree`)](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree).

I also note that [pre-commit (`elagil.pre-commit-helper`)](https://marketplace.visualstudio.com/items?itemName=elagil.pre-commit-helper) exists (apparently to invoke pre-commit on specific files or all files from within VS Code) but I don't currently use it. Unlike the Ruff extension, this appears “unofficial” in the sense that it's not developed or promoted by the developer of pre-commit.
